pyinduct/tests/test_core.py

```
@@ -278,6 +278,54 @@ def test_scalar_product_hint(self):
         g = pi.Function(np.sin, domain=(-17, 10))
         self.assertEqual(g.scalar_product_hint(), f.scalar_product_hint())
 
+    def test_apply_operator(self):
+        def linear_op(f):
+            return f * 2
+
+        def nonlinear_op(f):
+            return np.sqrt(f)
+
+        f = pi.Function(lambda z: np.exp(2 * z), domain=(0, 7),
+                        derivative_handles=[
+                            lambda z: 2 * np.exp(2 * z),
+                            lambda z: 4 * np.exp(2 * z)])
+        z_test = np.linspace(0, 7)
+
+        f_lin = f._apply_operator(linear_op, linear=True)
+        np.testing.assert_array_almost_equal(f_lin(z_test), 2 * f(z_test))
+        np.testing.assert_array_almost_equal(f_lin.derive(2)(z_test),
+                                             2 * f.derive(2)(z_test))
+
+        f_nli = f._apply_operator(nonlinear_op)
+        np.testing.assert_array_almost_equal(f_nli(z_test), np.sqrt(f(z_test)))
+        self.assertEqual(len(f_nli.derivative_handles), 0)
+
+    def test_real_imag_conj(self):
+        lam = 2 + 3j
+        lam_r = np.real(lam)
+        lam_i = np.imag(lam)
+        f = pi.Function(lambda z: np.exp(lam * z), domain=(0, 7),
+                        derivative_handles=[
+                            lambda z: lam * np.exp(lam * z),
+                            lambda z: lam ** 2 * np.exp(lam * z)])
+        fr = f.real()
+        fr_d = lambda z: np.exp(lam_r*z) * np.cos(lam_i*z)
+        d_fr_d = lambda z: np.real(lam * np.exp(lam * z))
+        dd_fr_d = lambda z: np.real(lam ** 2 * np.exp(lam * z))
+        fi = f.imag()
+        fi_d = lambda z: np.exp(lam_r*z) * np.sin(lam_i*z)
+        fc = f.conj()
```

Application of `conj`could be tested with a call to `scale` with a complex argument. That should surface the error.

---

not needed, test should be ok with the restriction to additivity